assets/js/components/UserMenu.js

```diff
@@ -0,0 +1,183 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * WordPress dependencies
+ */
+import { Fragment, useState, useRef, useEffect, useCallback } from '@wordpress/element';
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import Data from 'googlesitekit-data';
+import { getSiteKitAdminURL, clearWebStorage } from '../util';
+import Dialog from './dialog';
+import Button from './button';
+import Menu from './menu';
+import Modal from './modal';
+import { STORE_NAME as CORE_SITE } from '../googlesitekit/datastore/site/constants';
+import { STORE_NAME as CORE_USER } from '../googlesitekit/datastore/user/constants';
+
+const { useSelect } = Data;
+
+function UserMenu() {
+	const proxyPermissionsURL = useSelect( ( select ) => select( CORE_SITE ).getProxyPermissionsURL() );
+	const userEmail = useSelect( ( select ) => select( CORE_USER ).getEmail() );
+	const userPicture = useSelect( ( select ) => select( CORE_USER ).getPicture() );
+	const [ dialogActive, toggleDialog ] = useState( false );
+	const [ menuOpen, toggleMenu ] = useState( false );
+	const menuButtonRef = useRef();
+	const menuRef = useRef();
+
+	useEffect( () => {
+		global.addEventListener( 'mouseup', handleMenuClose );
+		global.addEventListener( 'keyup', handleMenuClose );
+		global.addEventListener( 'keyup', handleDialogClose );
+
+		return () => {
+			global.removeEventListener( 'mouseup', handleMenuClose );
+			global.removeEventListener( 'keyup', handleMenuClose );
+			global.removeEventListener( 'keyup', handleDialogClose );
+		};
+	}, [] );
+
+	const handleMenu = useCallback( () => {
+		toggleMenu( ! menuOpen );
+	}, [ menuOpen ] );
+
+	const handleMenuClose = ( e ) => {
+		if (
+			( ( 'keyup' === e.type && 27 === e.keyCode ) || 'mouseup' === e.type ) &&
+			! menuButtonRef.current.buttonRef.current.contains( e.target ) &&
+			! menuRef.current.menuRef.current.contains( e.target )
+		) {
+			toggleMenu( false );
+		}
+	};
+
+	const handleDialog = useCallback( () => {
+		toggleDialog( ! dialogActive );
+		toggleMenu( false );
+	}, [ dialogActive ] );
+
+	const handleMenuItemSelect = useCallback( ( index, e ) => {
+		if (
+			( ( 'keydown' === e.type && (
+				13 === e.keyCode || // Enter
+					32 === e.keyCode // Space
+			) ) ||
+				'click' === e.type // Mouse
+			) ) {
+			switch ( index ) {
+				case 0:
+					handleDialog();
+					break;
+				case 1:
+					if ( ! proxyPermissionsURL ) {
+						return;
+					}
+					global.location.assign( proxyPermissionsURL );
+					break;
+				default:
+					handleMenu();
+			}
+		}
+	}, [ proxyPermissionsURL, handleMenu, handleDialog ] );
+
+	const handleDialogClose = ( e ) => {
```

Comment:
Let's use `useCallback` for this function to memoize it.

Reply:
IMHO As explained in this post, using `useCallback` here with `[]` dependency is not worth it and will be less efficient than it already is.
If you still feel that a `useCallback` will be a better thing to do here, I will be happy to make the change 😊

Reply:
@kostyalmm i am not concerned about performance here. The reason why I think we need to use `useCallback` here is to make sure that we always have the same reference to the same `handleDialogClose` function, so `removeEventListener` calls in the cleanup callback can properly remove event listeners defined on the component mount.

Reply:
Ahh got it. Thanks, making the change